### ISSUE TITLE
loading StarDist3D models after training

### ISSUE THREAD

**Issue by xies** [mpicbg-csbd/stardist#35]
Hi! I've been trying to get some StarDist3D models trained. I used some CPU clusters to do a initial training test, which finished OK with the following output:

```
100/100 [==============================] - 826s 8s/step - loss: 1.3955 - prob_loss: 0.4074 - dist_loss: 4.9404 - prob_kld: 0.1733 - dist_relevant_mae: 4.9394 - dist_relevant_mse: 56.6804 - val_loss: 1.4757 - val_prob_loss: 0.3504 - val_dist_loss: 5.6265 - val_prob_kld: 0.1715 - val_dist_relevant_mae: 5.6254 - val_dist_relevant_mse: 66.9819
NMS threshold = 0.3:  75%|███████▌  | 15/20 [07:27<02:29, 29.85s/it, 0.513 -> 0.014]
NMS threshold = 0.4:  75%|███████▌  | 15/20 [08:30<02:50, 34.02s/it, 0.513 -> 0.009]
NMS threshold = 0.5:  75%|███████▌  | 15/20 [10:14<03:24, 41.00s/it, 0.513 -> 0.008]

Loading network weights from 'weights_best.h5'.
GPU enabled False: 16467.110308 sec
Using optimized values: prob_thresh=0.514525, nms_thresh=0.3.
Saving to 'thresholds.json'.
```

I try to subsequently reload this trained model onto an interactive node / my personal local machine with:

`model = StarDist3D(None, name='fixed_organoids_3d',basedir='/path/to/models')`

Two strange things happen. On the interactive node, the entire python shell crashes, which i suspect is because the CUDA-enabled tensorflow is not configured properly on the node:

 ```
Loading network weights from 'weights_best.h5'.
2020-02-18 07:25:19.385558: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
Aborted
```

But on my personal machine, where tensorflow-gpu is working OK (via gputools), I get the following error:

```
Traceback (most recent call last):

  File "<ipython-input-11-2244447c3551>", line 1, in <module>
    model = StarDist3D(None, name='fixed_organoids_2d')

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/stardist/models/model3d.py", line 294, in __init__
    super().__init__(config, name=name, basedir=basedir)

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/stardist/models/base.py", line 145, in __init__
    super().__init__(config=config, name=name, basedir=basedir)

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/csbdeep/models/base_model.py", line 92, in __init__
    self._set_logdir()

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/csbdeep/models/base_model.py", line 30, in wrapper
    return f(*args, **kwargs)

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/csbdeep/models/base_model.py", line 126, in _set_logdir
    self.config = self._config_class(**config_dict)

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/stardist/models/model3d.py", line 256, in __init__
    self.update_parameters(False, **kwargs)

  File "/anaconda2/envs/python3/lib/python3.6/site-packages/csbdeep/models/config.py", line 72, in update_parameters
    raise AttributeError("Not allowed to add new parameters (%s)" % ', '.join(attr_new))

AttributeError: Not allowed to add new parameters (train_foreground_only)
```

I'm really not sure what is going on since I'm not passing any new configuration to the model constructor.

Also, just wanted to add that some of the pre-trained 2D models are working excellently on my hard-to-segment 2d images! :)

**Comment by uschmidt83** [comment]
> NMS threshold = 0.3:  75%|███████▌  | 15/20 [07:27<02:29, 29.85s/it, 0.513 -> 0.014]
> NMS threshold = 0.4:  75%|███████▌  | 15/20 [08:30<02:50, 34.02s/it, 0.513 -> 0.009]
> NMS threshold = 0.5:  75%|███████▌  | 15/20 [10:14<03:24, 41.00s/it, 0.513 -> 0.008]

It is strange that the scores are essentially zero (0.014, 0.009, 0.008), they should be > 0.5 in most cases. Can you post an example training image pair (input/labels)?

> But on my personal machine, where tensorflow-gpu is working OK (via gputools)

Btw, [gputools](https://github.com/maweigert/gputools) is using OpenCL and has nothing to with TensorFlow.

> ```
> AttributeError: Not allowed to add new parameters (train_foreground_only)
> ```

I think that is because you're running the latest version of StarDist on your cluster, but an older version on your personal machine. It should work after you've upgraded.

> Also, just wanted to add that some of the pre-trained 2D models are working excellently on my hard-to-segment 2d images! :)

Great to hear!

Best,
Uwe

**Comment by xies** [comment]
Hi!

Thanks,  I made sure now that all the stardist modules are the same versions. Unfortunately, tensorflow is still not configured properly on the computing cluster, crashing with errors that say:
```
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
Aborted
```

with while trying to do very simple print statements, so this is something I need to solve with our sysadmin.

I am loading the trained module onto my local machine (OSX, Python3.5) and using it to predict on similar images. I can now load the cluster-trained (stardist 0.41) model onto my machine:
```
model = StarDist3D(None, name='modelname',
                   basedir=path.join(trainingdir,'models'))
labels, details = model.predict_instances(test_img[)
```

The kernel runs for a minute or so and then silently crashes without any error message, and then restarts silently (no warning messages on the bash shell that spyder is running in either).

<img width="804" alt="Screen Shot 2020-02-21 at 3 09 51 PM" src="https://user-images.githubusercontent.com/5126258/75041128-3cde2e00-54bc-11ea-9233-cd8212b14a3b.png">

--

The training/labels are 3D images of clusters of cells. One mid-level slice looks like this, with 1µm slices, about 20µm:

<img width="851" alt="Screen Shot 2020-02-21 at 3 02 38 PM" src="https://user-images.githubusercontent.com/5126258/75040633-4c10ac00-54bb-11ea-8f3a-f853a9fda51b.png">

**Comment by maweigert** [comment]
Hi @xies 

As @uschmidt83 already mentioned, your models seems to be not trained to something reasonable:

> It is strange that the scores are essentially zero (0.014, 0.009, 0.008), they should be > 0.5 in most cases. Can you post an example training image pair (input/labels)?

Did you normalize the input images properly? Can you send us a screenshot of the training/validation loss curves from tensorboard?

**Comment by xies** [comment]
Here are the tensorboard outputs:
<img width="393" alt="Screen Shot 2020-02-21 at 3 44 00 PM" src="https://user-images.githubusercontent.com/5126258/75044285-aad92400-54c1-11ea-8a8d-a1b0cd1b25a7.png">
<img width="397" alt="Screen Shot 2020-02-21 at 3 48 53 PM" src="https://user-images.githubusercontent.com/5126258/75044294-ae6cab00-54c1-11ea-8c42-348ee96984d1.png">

And the images:
<img width="455" alt="Screen Shot 2020-02-21 at 3 48 08 PM" src="https://user-images.githubusercontent.com/5126258/75045655-fee50800-54c3-11ea-81fa-1c0433697e77.png">
<img width="456" alt="Screen Shot 2020-02-21 at 3 48 11 PM" src="https://user-images.githubusercontent.com/5126258/75045657-ff7d9e80-54c3-11ea-8ee7-bbd94012c47b.png">

--

The wonky training is likely because in a hurry to iron out any installation problems (e.g. with tensorflow) I only generated two training datasets (since they take a while to complete), and since the random training/validation complained at me, I duplicated one of the training sets to feed to the training. I'm running a training now with one more training dataset.

Would this be related to the silent crashing on prediction? If the training outcomes are strange, wouldn't it still produce some model that will yield bad predictions (e.g. no objects or bad objects) instead of the crashing behavior?

**Comment by maweigert** [comment]
> Would this be related to the silent crashing on prediction?

Most likely yes: Judging from your training losses (dist_relevant_mae ist typically ~ 1) and network output (e.g. net_output0 looks bad) the training didn't succeed and thus at prediction time many strange objects may be proposed that kill the memory etc. I would change (in order of importance)

* Train longer (~100 epochs)
* Use more trainings images (~10)
* Use `grid = (2,2)` in  `Config2D()`

Hope that helps

**Comment by xies** [comment]
Thanks for the help! I am retraining with slightly more training images (~4), and longer training (50). I will report back when everything is done.

**Comment by xies** [comment]
Increasing training set + epochs have made the silent crash go away. Thanks for all your help!

**Comment by mballari** [comment]
I am having a similar issue in that my prediction scores are essentially 0, as shown in the first image. I trained over 400 epochs, using 54 training images, and used grid = (2,2) in Config2D( ). What could be a reason my model is being trained poorly? I have also attached an example training image pair. 

![image](https://user-images.githubusercontent.com/59752040/80570438-9fb5df00-89af-11ea-8917-156b973fcafc.png)

![image](https://user-images.githubusercontent.com/59752040/80571301-3c2cb100-89b1-11ea-91b4-afeea9a9d229.png)
